File: prset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// Use abstraction on all methods that sum the digit?
// - Differences: Starting point (from right), skip-interval, factor to multiply digits?

int getFirstSum(long cardNumber);
int getSecondSum(long cardNumber);
int sumDigits(int number);
int cardLength(long cardNumber);
void printCardType(long cardNumber, int cardLength);
int digit(long number, long position);
int digit_v2(long number, long position);

int main(void)
{
    long cardNumber = get_long("Number: ");
    // 1. Determine sum of the digtis of [every other digit (starting at second to rightmost) x 2]
    int firstSum = getFirstSum(cardNumber);

    // 2. Determine sum of every other digit (starting at rightmost) x 1
    int secondSum = getSecondSum(cardNumber);

    // 3. Sum "firstSum" and "secondSum"
    int checkSum = firstSum + secondSum;

    // Can we use modulo operator here?
    bool validCheckSum = (checkSum % 10 == 0);

    if (validCheckSum)
    {
        // Since the number is valid, we can determine its type
        printCardType(cardNumber, cardLength(cardNumber));
    }
    else
    {
        printf("INVALID\n");
    }
}

// getFirstSum
// Determines the sum of every other digit, starting at the second to rightmost
int getFirstSum(long cardNumber)
{

    // Current position of the digit we are searching for
    long place = 10;
    int sum = 0;

    while (place <= cardNumber)
    {
        sum += sumDigits(digit_v2(cardNumber, place) * 2);
        place *= 100;
    }

    return sum;


}

// getSecondSum
// Determines the sum of every other digit, starting at the rightmost
int getSecondSum(long cardNumber)
{
    // Current position of the digit we are searching for
    long place = 1;
    int sum = 0;

    while (place <= cardNumber)
    {
        sum += digit_v2(cardNumber, place);
        place *= 100;
    }

    return sum;
}

// sumDigits
int sumDigits(int number)
{
    int place = 1;
    int sum = 0;

    while (place <= number)
    {
        sum += digit_v2(number, place);
        place *= 10;
    }
    return sum;
}

// cardLength
// Determines the number of digits present in _cardNumber_
int cardLength(long cardNumber)
{
    long place = 1;
    int length = 0;
    while (place <= cardNumber)
    {
        place *= 10;
        length++;
    }
    return length;
}

// printCardType
// Determines the card type of a  valid _cardNumber_ with length _cardLength_
void printCardType(long cardNumber, int cardLength)
{
    int digit_16 = digit_v2(cardNumber, 1000000000000000);
    int digit_15 = digit_v2(cardNumber, 100000000000000);
    int digit_14 = digit_v2(cardNumber, 10000000000000);
    int digit_13 = digit_v2(cardNumber, 1000000000000);

    // American Express
    bool amex = (cardLength == 15 && digit_15 == 3 && (digit_14 == 4 || digit_14 == 7));
    // Put this into another variable for styling purposes
    bool mastercardSecondDigit = (digit_15 == 1 || digit_15 == 2 || digit_15 == 3 || digit_15 == 4 || digit_15 == 5);
    bool mastercard = cardLength == 16 && digit_16 == 5 && mastercardSecondDigit;
    bool visa = (cardLength == 13 && digit_13 == 4) || (cardLength == 16 && digit_16 == 4);

    if (amex)
    {
        printf("AMEX\n");
    }
    else if (mastercard)
    {
        printf("MASTERCARD\n");
    }
    else if (visa)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// digit
// Determines the digit in _place_ of _number_ without using the modulo operator

int digit(long number, long place)
{
    // The card number rounded to the digits place we are searching for
    long rounded = (number / (place * 10)) * (place * 10);

    long extra = number - rounded;

    // The digit we are looking for is the FIRST DIGIT of the difference of the extra discarded from rounding
    int digit = extra / place;

    return digit;
}

// digit_v2
// Determines the digit in _place_ of _number_ using the modulo operator

int digit_v2(long number, long place)
{
    // Plan: Determine the remainder when divided by the right place
    long extra = number % (place * 10);
    int digit = extra / place;
    return digit;
}

File: pset6/credit/credit.py
from cs50 import get_string
import re

# Ideas: Use regex to validate user input (correct length, only digits?)
# Criteria for valid card number:
# - Only includes numbers
# - Length is 13, 15, or 16
# - Proper checksum
# Easier to work with strings than numbers in this scenario (perform less arithmetic)


def main():

    user_input = get_string("Number: ")

    # Check for correct length and only diigts
    if (not(possible_card(user_input))):
        print("INVALID")
        return 1

    # Check if the check sums are valid
    if (not(valid_check_sum(user_input))):
        print("INVALID")
        return 2

    # Print the particular type of card
    print_card_type(user_input)

    return 0


# valid : String -> Boolean
# Is the provided user_input a possible card number?


def possible_card(user_input):
    match = re.fullmatch('[0-9]{13}|[0-9]{15}|[0-9]{16}', user_input)
    if match:
        return True
    return False


# valid_check_sum : String -> Boolean
# Do the digits in _card_number_ form a valid checksum?


def valid_check_sum(card_number):
    check_sum = first_sum(card_number) + second_sum(card_number)
    valid = check_sum % 10 == 0
    return valid


# first_sum : String -> Number
# Determines the sum of every other digit in _card_number_ times 2, starting at the second to rightmost


def first_sum(card_number):

    # Accumulator
    total = 0
    card_length = len(card_number)

    for i in range(1, card_length, 2):
        index = (card_length - 1) - i
        total += sum_digits(int(card_number[index]) * 2)

    return total


# sum_digits : Number -> Number
# Determines the sum of all digits in _num_


def sum_digits(num):

    # Accumulator
    total = 0
    digits = str(num)

    for i in range(0, len(digits)):
        total += int(digits[i])

    return total

# second_sum : String -> Number
# Determines the sum of every other digit in _card_number_, starting at the rightmost digit


def second_sum(card_number):

    # Accumulator
    total = 0
    card_length = len(card_number)

    for i in range(0, card_length, 2):
        index = (card_length - 1) - i
        total += int(card_number[index])

    return total


# print_card_type : String -> _
# Prints the type of card that _card_number_ represents, or "invalid"
# N.B. Card number consists of only digits


def print_card_type(card_number):
    # Use regex to determine if digits are in correct positions

    # AMEX: Contains 15 digits; first 2 are 34 or 37
    amex = re.fullmatch("34\d{13}|37\d{13}", card_number)
    mastercard = re.fullmatch("51\d{14}|52\d{14}|53\d{14}|54\d{14}|55\d{14}|", card_number)
    visa = re.fullmatch("4\d{12}|4\d{15}", card_number)

    if amex:
        print("AMEX")
    elif mastercard:
        print("MASTERCARD")
    elif visa:
        print("VISA")
    else:
        print("INVALID")


main()